pkg/authorization/authorizer/browsersafe/authorizer.go

```diff
@@ -0,0 +1,75 @@
+package browsersafe
+
+import (
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+)
+
+const (
+	proxyAction = "proxy"
+	unsafeProxy = "unsafeproxy"
+)
+
+type browserSafeAuthorizer struct {
+	delegate authorizer.Authorizer
+
+	// list of groups, any of which indicate the request is authenticated
+	authenticatedGroups sets.String
+}
+
+func NewBrowserSafeAuthorizer(delegate authorizer.Authorizer, authenticatedGroups ...string) authorizer.Authorizer {
+	return &browserSafeAuthorizer{
+		delegate:            delegate,
+		authenticatedGroups: sets.NewString(authenticatedGroups...),
+	}
+}
+
+func (a *browserSafeAuthorizer) Authorize(attributes authorizer.Attributes) (bool, string, error) {
+	browserSafeAttributes := a.getBrowserSafeAttributes(attributes)
+	return a.delegate.Authorize(browserSafeAttributes)
+}
+
+func (a *browserSafeAuthorizer) getBrowserSafeAttributes(attributes authorizer.Attributes) authorizer.Attributes {
+	if !attributes.IsResourceRequest() {
+		return attributes
+	}
+
+	isProxyVerb := attributes.GetVerb() == proxyAction
+	isProxySubresource := attributes.GetSubresource() == proxyAction
+
+	if !isProxyVerb && !isProxySubresource {
+		// Requests to non-proxy resources don't expose HTML or HTTP-handling user content to browsers
+		return attributes
+	}
+
+	if user := attributes.GetUser(); user != nil {
+		if a.authenticatedGroups.HasAny(user.GetGroups()...) {
+			// An authenticated request indicates this isn't a browser page load.
+			// Browsers cannot make direct authenticated requests.
+			// This depends on the API not enabling basic or cookie-based auth.
+			return attributes
+		}
+	}
+
+	browserSafeAttributes := authorizer.AttributesRecord{
```

Comment:
This is actually a good use case for inlining, and just implementing GetVerb and GetSubresource. We don't want to have to update this if new attributes are added

Reply:
> This is actually a good use case for inlining, and just implementing GetVerb and GetSubresource. We don't want to have to update this if new attributes are added

This is still outstanding.